### ISSUE TITLE
Error in plugin when upgrading DALI from v1.26 to v1.27

### ISSUE THREAD

**Issue by fversaci** [NVIDIA/DALI#5033]
### Version

1.27

### Describe the bug.

Hi,

I have tested our Cassandra plugin with DALI v1.27 and we came across an "Index out of range" error. This error is not present when using DALI v1.26. The program runs without any errors, but crashes only when it is being exited. Do you have any idea which specific change from v1.26 to v1.27 might be causing it and if there is a potential fix for it?

### Minimum reproducible example

```shell
git clone https://github.com/fversaci/cassandra-dali-plugin.git
cd cassandra-dali-plugin
git checkout bug-dali127
docker build -t cassandra-dali-plugin .   # this might take some time
docker run --cap-add=sys_admin --rm -it --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --name cass-dali cassandra-dali-plugin
# within the container
cd examples/imagenette/
./start-and-fill-db.sh 
python3 loop_read.py --epochs 1   # crashes when exiting
```

### Relevant log output

```shell
terminate called after throwing an instance of 'dali::DALIException'
  what():  [/opt/dali/dali/pipeline/graph/op_graph.h:197] Index -1 out of range [0, 3).
Stacktrace (36 entries):
[frame 0]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/libdali.so(+0xd2c0b) [0x7fdaab069c0b]
[frame 1]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/libdali.so(dali::OpGraph::Node(long)+0x1b8) [0x7fdaab0acca8]
[frame 2]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/libdali.so(dali::Pipeline::~Pipeline()+0xcc) [0x7fdaab15c8ec]
[frame 3]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/backend_impl.cpython-310-x86_64-linux-gnu.so(+0x6d7a5) [0x7fdaa0c927a5]
[frame 4]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/python_function_plugin.cpython-310-x86_64-linux-gnu.so(+0x37d50) [0x7fdaab96fd50]
[frame 5]: /usr/local/lib/python3.10/dist-packages/nvidia/dali/python_function_plugin.cpython-310-x86_64-linux-gnu.so(+0x388d1) [0x7fdaab9708d1]
[frame 6]: python3(+0x13be43) [0x562e32723e43]
[frame 7]: python3(+0x178390) [0x562e32760390]
[frame 8]: python3(+0x1561d0) [0x562e3273e1d0]
[frame 9]: python3(_PyFunction_Vectorcall+0xb3) [0x562e32748743]
[frame 10]: python3(PyObject_Call+0x122) [0x562e32757192]
[frame 11]: python3(_PyEval_EvalFrameDefault+0x2b71) [0x562e327332c1]
[frame 12]: python3(_PyObject_FastCallDictTstate+0xc4) [0x562e3273d784]
[frame 13]: python3(_PyObject_Call_Prepend+0x5c) [0x562e3275354c]
[frame 14]: python3(+0x2841e0) [0x562e3286c1e0]
[frame 15]: python3(PyObject_Call+0xbb) [0x562e3275712b]
[frame 16]: python3(_PyEval_EvalFrameDefault+0x2b71) [0x562e327332c1]
[frame 17]: python3(_PyFunction_Vectorcall+0x7c) [0x562e3274870c]
[frame 18]: python3(_PyEval_EvalFrameDefault+0x2b71) [0x562e327332c1]
[frame 19]: python3(_PyObject_FastCallDictTstate+0xc4) [0x562e3273d784]
[frame 20]: python3(_PyObject_Call_Prepend+0x5c) [0x562e3275354c]
[frame 21]: python3(+0x2841e0) [0x562e3286c1e0]
[frame 22]: python3(_PyObject_MakeTpCall+0x25b) [0x562e3273e5eb]
[frame 23]: python3(_PyEval_EvalFrameDefault+0x6516) [0x562e32736c66]
[frame 24]: python3(+0x239e56) [0x562e32821e56]
[frame 25]: python3(PyEval_EvalCode+0x86) [0x562e32821cf6]
[frame 26]: python3(+0x2647d8) [0x562e3284c7d8]
[frame 27]: python3(+0x25e0bb) [0x562e328460bb]
[frame 28]: python3(+0x264525) [0x562e3284c525]
[frame 29]: python3(_PyRun_SimpleFileObject+0x1a8) [0x562e3284ba08]
[frame 30]: python3(_PyRun_AnyFileObject+0x43) [0x562e3284b653]
[frame 31]: python3(Py_RunMain+0x2be) [0x562e3283e41e]
[frame 32]: python3(Py_BytesMain+0x2d) [0x562e32814cad]
[frame 33]: /usr/lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fdb2af1fd90]
[frame 34]: /usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7fdb2af1fe40]
[frame 35]: python3(_start+0x25) [0x562e32814ba5]
```

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @fversaci,

Thank you for reaching out. It seems that this may be caused by https://github.com/NVIDIA/DALI/pull/4874 which traverses the operator's graph during the pipeline teardown. However it is hard to tell if this is the cause or of it just exposed the bug that was already there,
@klecki any idea why we may not find a valid tensor with a given name - https://github.com/NVIDIA/DALI/blame/release_v1.27/dali/pipeline/pipeline.cc#L156 ?

**Comment by JanuszL** [comment]
I have an idea. Let me see if https://github.com/NVIDIA/DALI/pull/5036 helps.

**Comment by JanuszL** [comment]
Yes, I have tested the build with the mentioned change and it helps. I hope to merge it soon so you can test the nightly build that follows.

**Comment by fversaci** [comment]
Thanks @JanuszL, looking forward to it!

**Comment by JanuszL** [comment]
@fversaci that is an excellent question.
I checked more thoroughly and it seems to be a problem on my side.
There is a logic used to feed input operators that matches the name of the operator to its instance in the graph. The logic assumes there is only one output. In the case of multiple outputs, the naming schema is different ('Reader[0]' instead of just `Reader`).
I copied that logic to the place executed during pipeline shutdown, and the Casandra operator which has two outputs escaped that. The fix was to abstract away from the naming schema during the pipeline teardown.
